packages/transform/src/diagnostics/augmentation.ts

```diff
@@ -0,0 +1,146 @@
+  mapping: MappingTree
+): ts.DiagnosticMessageChain | undefined {
+  let node = mapping.sourceNode;
+  let parentNode = mapping.parent?.sourceNode;
+  if (!parentNode) return;
+
+  if (
+    node.type === 'Identifier' &&
+    parentNode.type === 'AttrNode' &&
+    !/^(@|\.)/.test(parentNode.name)
+  ) {
+    // If the assignability issue is on an attribute name and it's not an `@arg`
+    // or `...attributes`, then it's an HTML attribute type issue.
+    return addGlintDetails(
+      message,
+      'Only primitive values (see `AttrValue` in `@glint/template`) are assignable as HTML attributes. ' +
+        'If you want to set an event listener, consider using the `{{on}}` modifier instead.'
+    );
+  } else if (
+    node.type === 'MustacheStatement' &&
+    (parentNode.type === 'Template' ||
+      parentNode.type === 'BlockStatement' ||
+      parentNode.type === 'ElementNode')
+  ) {
+    // Otherwise, if it's on a full {{mustache}} and it's in a top-level position,
+    // it's a DOM content type issue.
+    return addGlintDetails(
+      message,
+      'Only primitive values and certain DOM objects (see `ContentValue` in `@glint/template`) are ' +
+        'usable as top-level template content.'
+    );
+  }
+}
+
+function adjustArgumentArity(
+  diagnostic: Diagnostic,
+  mapping: MappingTree
+): string | ts.DiagnosticMessageChain {
+  // For normal template entity invocations, named args count as a function argument,
+  // so TS reports one more required and passed positional arg than is actual visible
+  // in the template. For `{{yield}}`, there are two extra args (𝚪 and the block name).
+  //
+  // Since (at least in English) TS always says "arguments" regardless of the
+  // actual plurality, we can just blindly subtract 1 (or 2 for yields) from all numeric
+  // patterns in the message.
+  let callNode = findAncestor(mapping, 'MustacheStatement', 'SubExpression');
+  let delta = callNode?.path.loc.asString() === 'yield' ? 2 : 1;
+
+  return mapMessageText(diagnostic.messageText, (message) =>
+    message.replace(/\d+/g, (n) => `${Number(n) - delta}`)
+  );
+}
+
+function mapMessageText<T extends string | ts.DiagnosticMessageChain>(
+  message: T,
+  mapper: (text: string) => string
+): T {
+  if (typeof message === 'string') {
+    return mapper(message) as T;
+  } else {
+    return {
+      ...(message as ts.DiagnosticMessageChain),
+      messageText: mapper(message.messageText),
+      next: message.next?.map((next) => mapMessageText(next, mapper)),
+    } as T;
+  }
+}
+
+function checkResolveError(
+  diagnostic: Diagnostic,
+  mapping: MappingTree
+): ts.DiagnosticMessageChain | undefined {
+  let nodeType = mapping.sourceNode.type;
+  if (nodeType === 'ElementNode' || nodeType === 'PathExpression' || nodeType === 'Identifier') {
+    return addGlintDetails(
+      diagnostic,
+      'The given value is not resolvable as an entity that can be invoked in a template, ' +
+        'such as a component, modifier or helper.'
```

Comment:
I don't *yet* have a great suggestion for it, but I wonder if we can wordsmith this a bit to be a bit clearer for people who don't have our level of context. 🤔 "resolvable as an entity that can be invoked" *feels* deep in the weeds. Hmmmmm.

Reply:
I had a similar feeling (and meant to call out in the PR description that phrasing suggestions on all these messages are welcome!) but wasn't immediately able to come up with anything clearer that wasn't somehow misleading.

Reply:
Maybe just `The given value is not resolvable as a component, modifier or helper` is better?